File: src/leetCode/c++/Solution435.cpp
//Intervals O(nlogn) O(n)
class Solution435 {
public:
    int eraseOverlapIntervals(vector<vector<int>>& intervals) {
        sort(intervals.begin(), intervals.end());
        int right = INT_MIN, res = 0;
        for(auto interval : intervals){
            if(interval[0] < right){//前一個的右側大於這左側
                res++;
                right = min(right, interval[1]);//取較短的, 即捨棄長的
            }else{
                right = interval[1];
            }
        }
        return res;
    }
};

File: java/Solution435.java
package leetCode.java;

import java.util.*;

//Intervals O(nlogn) O(n)
class Solution435 {
    public int eraseOverlapIntervals(int[][] intervals) {
        // 盡量多保留interval, 回傳最少要移除的量
        Arrays.sort(intervals, (a, b) -> a[1] - b[1]); // 依照end小排到大, 可以最緊湊
        int end = Integer.MIN_VALUE, res = 0;
        for (int[] interval : intervals) {
            if (interval[0] < end) { // 左界在上一個的右界內, 有交集, 移除
                ++res;
                continue;
            }
            end = interval[1]; // 更新最右側
        }

        return res;
    }
}